src/App.tsx

```diff
@@ -1,26 +1,223 @@
+  const [shouldDeleteCompleted, setShouldDeleteCompleted] = useState(false);
+  const [newTodo, setNewTodo] = useState<Todo | null>(null);
+  const [todoBeingAdded, setTodoBeingAdded] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [clearInput, setClearInput] = useState(false);
+  const focusedTodoRef = useRef<HTMLInputElement>(null);
+  const defaultInputRef = useRef<HTMLInputElement>(null);
+
+  const showError = (errMessage: string) => {
+    if (errMessage) {
+      setErrorMessage(errMessage);
+      setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+    }
+  };
+
+  const getCompletedTodos = React.useCallback(() => {
+    return currentTodos.filter(todo => todo.completed);
+  }, [currentTodos]);
+
+  const deleteChosenTodo = React.useCallback(
+    (currentTodoToDelete: Todo | null) => {
+      if (currentTodoToDelete) {
+        deleteTodo(currentTodoToDelete)
+          .then(() => {
+            setCurrentTodos([
+              ...currentTodos.slice(
+                0,
+                currentTodos.indexOf(currentTodoToDelete),
+              ),
+              ...currentTodos.slice(
+                currentTodos.indexOf(currentTodoToDelete) + 1,
+              ),
+            ]);
+          })
+          .catch(error => {
+            showError('Unable to delete a todo');
+            throw error;
+          })
+          .finally(() => {
+            setTodoToDelete(null);
+            defaultInputRef.current?.focus();
+          });
+      }
+    },
+    [currentTodos],
+  );
+
+  const getAndShowTodos = React.useCallback(() => {
+    getTodos()
+      .then(todos => {
+        setTodosFromServer(todos);
+        setCurrentTodos(todos);
+      })
+      .catch(error => {
+        showError('Unable to load todos');
+        throw error;
+      })
+      .finally(() => defaultInputRef.current?.focus());
+  }, []);
+
+  const postNewTodo = React.useCallback(
+    (todoToPost: Todo | null) => {
+      if (todoToPost) {
+        setTodoBeingAdded(true);
+        setTempTodo({
+          id: 0,
+          userId: 2400,
+          title: todoToPost.title,
+          completed: todoToPost.completed,
+        });
+
+        setErrorMessage('');
+
+        postTodo(todoToPost)
+          .then(() => {
+            setCurrentTodos([...currentTodos, todoToPost]);
+            setClearInput(true);
+            setNewTodo(null);
+          })
+          .catch(error => {
+            showError('Unable to add a todo');
+            throw error;
+          })
+          .finally(() => {
+            setTodoBeingAdded(false);
+            setTempTodo(null);
+          });
+      }
+    },
+    [currentTodos],
+  );
+
+  const deleteCompletedTodos = React.useCallback(() => {
+    if (shouldDeleteCompleted) {
+      const completedTodos = getCompletedTodos();
+
+      Promise.all(completedTodos.map(todo => deleteTodo(todo)))
```

Comment:
```suggestion
      Promise.allSettled(completedTodos.map(todo => deleteTodo(todo)))
```
Consider filtering todos with success response to update the array in the state, and fix the test